src/main/java/org/graylog/integrations/ipfix/codecs/IpfixCodec.java

```diff
@@ -0,0 +1,212 @@
+import javax.inject.Inject;
+import java.net.InetSocketAddress;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Codec(name = "ipfix", displayName = "IPFIX Codec")
+public class IpfixCodec extends AbstractCodec implements MultiMessageCodec {
+    private static final Logger LOG = LoggerFactory.getLogger(IpfixCodec.class);
+
+    private final IpfixAggregator ipfixAggregator;
+    private final IpfixParser parser;
+    private InformationElementDefinitions infoElementDefs;
+
+    @Inject
+    protected IpfixCodec(@Assisted Configuration configuration, IpfixAggregator ipfixAggregator) {
+        super(configuration);
+        this.ipfixAggregator = ipfixAggregator;
+        infoElementDefs = new InformationElementDefinitions(
+                Resources.getResource("ipfix-iana-elements.json"),
+                Resources.getResource("ixia-ipfix.json"));
+        this.parser = new IpfixParser(this.infoElementDefs);
+    }
+
+    /**
+     * Parses out the fields from the flow record, assigns v5 fixed format fields and create message
+     *
+     * @param record
+     * @return
+     */
+    private static String toMessageString(Flow record) {
+        LOG.debug("IPFIX message being assembled from flow record [{}].", record.fields());
+        final ImmutableMap<String, Object> fields = record.fields();
+        final long packetCount = (long) fields.getOrDefault("packetDeltaCount", 0L);
+        long octetCount = (long) fields.getOrDefault("octetDeltaCount", 0L);
+        if (octetCount == 0L) {
+            octetCount = (long) fields.getOrDefault("fwd_flow_delta_bytes", 0L);
+        }
+
+        String srcAddr = (String) fields.get("sourceIPv4Address");
+        String dstAddr = (String) fields.get("destinationIPv4Address");
+        if (srcAddr == null) {
+            srcAddr = (String) fields.get("sourceIPv6Address");
+        }
+        if (dstAddr == null) {
+            dstAddr = (String) fields.get("destinationIPv6Address");
+        }
+
+        final Number srcPort = (Number) fields.get("sourceTransportPort");
+        final Number dstPort = (Number) fields.get("destinationTransportPort");
+        final long protocol = Long.parseLong(String.valueOf(fields.getOrDefault("protocolIdentifier", 0L)));
+
+        // TODO should this be configurable?
+        return createMessageString(packetCount, octetCount, srcAddr, dstAddr, srcPort, dstPort, protocol);
+    }
+
+    private static String createMessageString(long packetCount, long octetCount, String srcAddr, String dstAddr,
+                                              Number srcPort, Number dstPort, long protocol) {
+        String message = String.format(Locale.ROOT, "Ipfix [" + srcAddr + "]:" + srcPort + " <> [" + dstAddr + "]:" + dstPort + " " +
```

Comment:
Do we need to split the srcAddr into two parts srcIPV4 and srcIPV6 . [https://whatismyipaddress.com/dual-stack](Dual Stack)

Reply:
No, for now we just need one value for both src and dest.